library/core/src/num/uint_macros.rs

```diff
@@ -1378,6 +1449,35 @@ macro_rules! uint_impl {
             (a as Self, b)
         }
 
+        /// Calculates `self` + `rhs` with a signed `rhs`
+        ///
+        /// Returns a tuple of the addition along with a boolean indicating
+        /// whether an arithmetic overflow would occur. If an overflow would
+        /// have occurred then the wrapped value is returned.
+        ///
+        /// # Examples
+        ///
+        /// Basic usage:
+        ///
+        /// ```
+        /// # #![feature(uint_add_signed)]
+        #[doc = concat!("assert_eq!(1", stringify!($SelfT), ".overflowing_add_signed(2), (3, false));")]
+        #[doc = concat!("assert_eq!(1", stringify!($SelfT), ".overflowing_add_signed(-2), (", stringify!($SelfT), "::MAX, true));")]
+        #[doc = concat!("assert_eq!((", stringify!($SelfT), "::MAX - 2).overflowing_add_signed(4), (1, true));")]
+        /// ```
+        #[unstable(feature = "uint_add_signed", issue = "none")]
+        #[rustc_const_unstable(feature = "uint_add_signed", issue = "none")]
+        #[must_use = "this returns the result of the operation, \
+                      without modifying the original"]
+        #[inline]
+        pub const fn overflowing_add_signed(self, rhs: $SignedT) -> (Self, bool) {
+            if rhs >= 0 {
+                self.overflowing_add(rhs as Self)
+            } else {
+                self.overflowing_sub(rhs.unsigned_abs())
+            }
```

Comment:
this is equivalent to the following (at least it passes all test values from https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=641a3ff738671aef5bd6f6bec5721010):

```suggestion
            let (res, overflowed) = self.overflowing_add(rhs as Self);
            (res, overflowed ^ (rhs < 0))
```

and this implementation gets rid of much of the conditional stuff:

```asm
# this PR
playground::overflowing_add_signed:
	mov	rax, rdi
	xor	r8d, r8d
	add	rdi, rsi
	setb	r8b
	mov	rcx, rsi
	neg	rcx
	xor	edx, edx
	sub	rax, rcx
	setb	dl
	test	rsi, rsi
	cmovns	rax, rdi
	cmovns	edx, r8d
	ret

# this review comment
playground::overflowing_add_signed_2:
	mov	rax, rdi
	add	rax, rsi
	setb	cl
	test	rsi, rsi
	sets	dl
	xor	dl, cl
	ret
```

Reply:
Thanks this is great !

On other archs it also gets rid of a branch. It also improves `checked_add_signed` which is implemented with it.